### PR TITLE
Add system of table views

### PR DESCRIPTION
**Pull request by zhongyuankai**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add a system table of table_views, convenient to query the dependency relationship between tables and views.

### CODE REVIEW

File: src/Storages/System/StorageSystemTableViews.cpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    for (const auto & [storage_id, view_ids] : view_dependencies)
+    {
+        const bool check_access_for_tables = check_access_for_databases && !access->isGranted(AccessType::SHOW_TABLES, storage_id.database_name);
+
+        if (check_access_for_tables && !access->isGranted(AccessType::SHOW_TABLES, storage_id.database_name, storage_id.table_name))
+            continue;
+
+        for (const auto & view_id : view_ids)
+        {
+            size_t col_num = 0;
```
Comment (kssenii) [inline src/Storages/System/StorageSystemTableViews.cpp:38]:
```suggestion
            size_t col_num = 0;
```
```suggestion
        size_t col_num;
        for (const auto & view_id : view_ids)
        {
            col_num = 0;
```

**Review by kssenii — APPROVED** [review]

File: src/Interpreters/DatabaseCatalog.cpp [hunk 2]
```diff
@@ -608,6 +608,12 @@ Dependencies DatabaseCatalog::getDependencies(const StorageID & from) const
         return {};
     return Dependencies(iter->second.begin(), iter->second.end());
 }
+ 
+void DatabaseCatalog::getViewDependencies(ViewDependencies & view_dependencies_) const
+{
+    std::lock_guard lock{databases_mutex};
+    view_dependencies_ = view_dependencies;
```
Comment (kitaisreal) [inline src/Interpreters/DatabaseCatalog.cpp:615]: It is not thread safe interface, we should copy view dependencies to client.
Better API will be:
```
ViewDependencies DatabaseCatalog::getViewDependencies()
```

**Review by kitaisreal — CHANGES_REQUESTED** [review]

**Comment by nikitamikhaylov** [comment]
Better to look at #25291 and #25714 to be familiar with current status of work near views. Also I noticed that there is a column called `dependencies_table` inside system.tables table. So, I think we don't need extra system tables. Maybe just add some new columns to existing tables. What do you think @Algunenano ?

File: src/Storages/System/StorageSystemTableViews.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace DB
+{
+
+class Context;
+
+NamesAndTypesList StorageSystemTableViews::getNamesAndTypes()
+{
+    return {
+        {"database", std::make_shared<DataTypeString>()},
```
Comment (kitaisreal) [inline src/Storages/System/StorageSystemTableViews.cpp:14]: I am not sure if naming is good enough.
From my perspective better name for table is system.views.
For columns table system.views should have same naming as system.tables.
`database` - database of view
`name` - name of view
`table` - view table
`table_database` - view table database.

**Comment by Algunenano** [comment]
> Better to look at #25291 and #25714 to be familiar with current status of work near views. Also I noticed that there is a column called dependencies_table inside system.tables table. So, I think we don't need extra system tables. Maybe just add some new columns to existing tables. What do you think @Algunenano ?

Yes, you can do the same as what's seen in the PR test with something like:

```sql
SELECT
    database,
    name
FROM system.tables
WHERE has(dependencies_table, 'table_views_test_view')
```

It's not very convenient since you should be joining the dependencies_database and dependencies_table arrays but it's doing the same. It could be more convenient.

**Comment by nikitamikhaylov** [comment]
My only point is that we add a code that doesn't provide a new functionality. The better way is to implement something like `system.views` with all information we could get (view type, its query, time of creation), but all this information seems to be in `system.tables` table.

**Comment by zhongyuankai** [comment]
I think that in scenarios where there are many views, it will be very convenient to have a system.views to query the dependencies between views and tables, and it does not conflict with the metadata management of system.tables.

**Comment by nikitamikhaylov** [comment]
@zhongyuankai Is it possible to add `Enum` column called `type`? View could be Plain, Materialized or Live.

**Comment by zhongyuankai** [comment]
> @zhongyuankai Is it possible to add `Enum` column called `type`? View could be Plain, Materialized or Live.

Yes, it will be more intuitive to add the `type` column

**Comment by Algunenano** [comment]
Note the views log currently has one type enum at https://github.com/ClickHouse/ClickHouse/blob/884b51116dc8c1ca12272c70f8d532e7c393f166/src/Interpreters/QueryViewsLog.h#L28 and https://github.com/ClickHouse/ClickHouse/blob/884b51116dc8c1ca12272c70f8d532e7c393f166/src/Interpreters/QueryViewsLog.cpp#L28 so it would be nice it that were put in common with the type here too.

**Comment by zhongyuankai** [comment]
@kitaisreal hi, please review again.

**Review by kssenii — APPROVED** [review]
LGTM

**Comment by sevirov** [comment]
Internal documentation ticket: DOCSUP-13927

File: src/Storages/System/StorageSystemViews.cpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+class Context;
+
+NamesAndTypesList StorageSystemViews::getNamesAndTypes()
+{
+    auto view_type_datatype = std::make_shared<DataTypeEnum8>(DataTypeEnum8::Values{
+        {"Default", static_cast<Int8>(QueryViewsLogElement::ViewType::DEFAULT)},
+        {"Materialized", static_cast<Int8>(QueryViewsLogElement::ViewType::MATERIALIZED)},
+        {"Live", static_cast<Int8>(QueryViewsLogElement::ViewType::LIVE)}});
+
+    return {
```
Comment (alexey-milovidov) [inline src/Storages/System/StorageSystemViews.cpp:22]: @kssenii We need better column names. Now they can be easily confused to each other. Also we need to highlight that `table` is just the main table in the sequence of JOINs.

Comment (kssenii) [reply src/Storages/System/StorageSystemViews.cpp:22]: At first column names were: `database`, `table`, `view_database`, `view_table`, but it was suggested be Maxim [in the comment above](https://github.com/ClickHouse/ClickHouse/pull/28082#discussion_r694887321) that they should be as they are now:
```
I am not sure if naming is good enough.
From my perspective better name for table is system.views.
For columns table system.views should have same naming as system.tables.
database - database of view
name - name of view
table - view table
table_database - view table database.
```

Comment (alexey-milovidov) [reply src/Storages/System/StorageSystemViews.cpp:22]: Ok.

Comment (alexey-milovidov) [reply src/Storages/System/StorageSystemViews.cpp:22]: What about:
`main_dependency_table`, `main_dependency_database`?

Comment (kitaisreal) [reply src/Storages/System/StorageSystemViews.cpp:22]: @alexey-milovidov I am not sure if this table system.views is actually needed, maybe better to add something to system.tables using Map data type.

**Pull request merged**